official/wasm/pkg/goanalysis/goanalysis.go

```diff
@@ -0,0 +1,246 @@
+					case *ast.ReturnStmt:
+						// If there is a return, insert before it.
+						funcDecl.Body.List = append(funcDecl.Body.List[:len(funcDecl.Body.List)-1], &ast.ExprStmt{X: insertionExpr}, lastStmt)
+					default:
+						// If there is no return, insert at the end of the function body.
+						funcDecl.Body.List = append(funcDecl.Body.List, &ast.ExprStmt{X: insertionExpr})
+					}
+				} else {
+					// If there are no statements in the function body, insert at the end of the function body.
+					funcDecl.Body.List = append(funcDecl.Body.List, &ast.ExprStmt{X: insertionExpr})
+				}
+				found = true
+				return false
+			}
+		}
+		return true
+	})
+
+	if !found {
+		return "", errors.Errorf("function %s not found", functionName)
+	}
+
+	// Write the modified AST to a buffer.
+	var buf bytes.Buffer
+	if err := format.Node(&buf, fileSet, f); err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// ReplaceReturn replaces return statements in a Go function with a new return statement.
+func ReplaceReturn(fileContent, functionName string, returnVars ...string) (string, error) {
+	fileSet := token.NewFileSet()
+
+	// Parse the Go source code content.
+	f, err := parser.ParseFile(fileSet, "", fileContent, parser.ParseComments)
+	if err != nil {
+		return "", err
+	}
+
+	returnStmts := make([]ast.Expr, 0)
+	for _, returnVar := range returnVars {
+		// Parse the new return var to expression.
+		newRetExpr, err := parser.ParseExpr(returnVar)
+		if err != nil {
+			return "", err
+		}
+		returnStmts = append(returnStmts, newRetExpr)
+	}
+
+	found := false
+	ast.Inspect(f, func(n ast.Node) bool {
+		if funcDecl, ok := n.(*ast.FuncDecl); ok {
+			// Check if the function has the name you want to replace.
+			if funcDecl.Name.Name == functionName {
+				// Replace the return statements.
+				for _, stmt := range funcDecl.Body.List {
+					if retStmt, ok := stmt.(*ast.ReturnStmt); ok {
+						// Remove existing return statements.
+						retStmt.Results = nil
+						// Add the new return statement.
+						retStmt.Results = append(retStmt.Results, returnStmts...)
+					}
+				}
+				found = true
+				return false
+			}
+		}
+		return true
+	})
+
+	if !found {
+		return "", errors.Errorf("function %s not found", functionName)
+	}
+
+	// Write the modified AST to a buffer.
+	var buf bytes.Buffer
+	if err := format.Node(&buf, fileSet, f); err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// ReplaceCode replaces a function implementation in Go source code content.
+func ReplaceCode(fileContent, oldFunctionName, newFunction string) (modifiedContent string, err error) {
+	fileSet := token.NewFileSet()
+
+	// Parse the Go source code content.
+	f, err := parser.ParseFile(fileSet, "", fileContent, parser.ParseComments)
+	if err != nil {
+		return "", err
+	}
+
+	// Parse the content of the new function into an ast.File.
+	newFuncContent := fmt.Sprintf("package p; func _() { %s }", strings.TrimSpace(newFunction))
```

Comment:
I'm not sure but wouldn't it be possible to just generate the `BlockStmt` instead of the function? Maybe using something like `{ %s }` to insert the function code

Reply:
This is the only way I found to parse different types of code; the `parser.ParseExpr` or `parser.ParseExprFrom` constantly crashes in other cases.